packages/lib/src/cli-config.ts

```diff
@@ -1,78 +1,111 @@
+	profileName: string
+}
 
-		if (this._config == null) {
-			const parsed = yaml.load(fs.readFileSync(`${this._configFile}`, 'utf-8'))
-			if (parsed) {
-				if (typeof parsed === 'object') {
-					this._config = { ...parsed }
-				} else {
-					throw new Error('invalid config file format; please specify zero or more profiles')
-				}
-			}
-		}
+export interface CLIConfig extends CLIConfigDescription {
+	profiles: ProfilesByName
+	managedProfiles: ProfilesByName
+	mergedProfiles: ProfilesByName
 
-		if (!this._config) {
-			// empty file; use empty map
-			this._config = {}
-		}
+	/**
+	 * A convenience reference to the selected profile (same as `mergedConfig[profileName]`).
+	 */
+	profile: Profile
+}
 
-		return this._config
+export const loadConfigFile = async (filename: string): Promise<ProfilesByName> => {
+	if (!yamlExists(filename)) {
+		return {}
 	}
 
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	public getRawConfigData(): { [name: string]: any } {
-		if (!this._config) {
-			throw new Error('config not initialized completely')
+	const parsed = cliConfigModule.load(await readFile(`${filename}`, 'utf-8'))
+	if (parsed) {
+		if (typeof parsed === 'object' && !Array.isArray(parsed)) {
+			const errors: string[] = []
+			const config: ProfilesByName = {}
+			for (const [profileName, profile] of Object.entries(parsed)) {
+				if (typeof(profile) === 'object' && !Array.isArray(profile)) {
+					config[profileName] = profile
+				} else {
+					errors.push(`bad profile ${profileName}; profile must be an object`)
+				}
+			}
+			if (errors.length) {
+				throw Error(`${errors.join('\n')}\n${seeConfigDocs}`)
+			}
+			return config
+		} else {
+			throw Error('invalid config file format\n' + seeConfigDocs)
 		}
-		return this._config
 	}
+	return {}
+}
 
-	public getProfile(name: string): Record<string, unknown> {
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		const config: { [name: string]: any } = this.loadConfig()
-		if (!(name in config)) {
-			return {}
-		}
-		const retVal = config[name]
-		if (retVal === null) {
-			throw new Error('null profile specified. Check config.yaml for errors.')
-		}
-		if (typeof retVal === 'object') {
-			return retVal
+/**
+ * Merge profiles from `preferred` and `secondary`, favoring any config entries from `preferred`
+ * over those in `secondary`.
+ */
+export const mergeProfiles = (preferred: ProfilesByName, secondary: ProfilesByName): ProfilesByName => {
+	const mergeInto = (into: ProfilesByName, from: ProfilesByName): ProfilesByName => {
+		for (const [profileName, profileData] of Object.entries(from)) {
+			if (!(profileName in into)) {
+				into[profileName] = {}
+			}
+			for (const [configKey, configValue] of Object.entries(profileData)) {
+				into[profileName][configKey] = configValue
+			}
 		}
-		throw new Error(`bad profile configuration for ${name} in ${this._configFile}`)
+		return into
 	}
+	return mergeInto(mergeInto({}, secondary), preferred)
```

Comment:
Is the nested call necessary?

Reply:
Not strictly, no, but it makes it so the inputs are unmodified which is always a good practice.